src/resources/primitive_kv_flash_esp32.cc

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2021 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "../top.h"
+
+#if defined(TOIT_FREERTOS)
+
+#include <nvs_flash.h>
+
+#include "../resource.h"
+#include "../objects_inline.h"
+#include "../process.h"
+
+namespace toit {
+const size_t MAX_KEY_LENGTH_ = 15;
+class PersistentResourceGroup : public ResourceGroup {
+ public:
+  TAG(PersistentResourceGroup);
+  PersistentResourceGroup(nvs_handle handle, Process* process)
+      : ResourceGroup(process, null)
+      , _handle(handle) {}
+
+  ~PersistentResourceGroup() {
+    nvs_close(_handle);
+  }
+
+  nvs_handle handle() { return _handle; }
+
+ private:
+  nvs_handle _handle;
+};
+
+bool is_valid_key(const char* key, Process* process) {
+  if (key[0] == '\0' || strlen(key) > MAX_KEY_LENGTH_) return false;
+  if (!process->is_privileged() && key[0] == '_') return false;
```

Comment:
I think this check should be present in the POSIX/Windows version too. Maybe add an `is_valid_key` helper there too?